### PR TITLE
feat: ability to set registry mirrors

### LINKED ISSUE #37: Feature request: support the ability to configure a registry mirror
Thanks for your work on `oci-build-task`! Is it worth exposing to users the ability to configure a [registry mirror with a pull-through cache](https://docs.docker.com/registry/recipes/mirror/)?

## Why would this be helpful?

Docker Hub will begin [exposing strict rate limits November 1](https://www.docker.com/blog/scaling-docker-to-serve-millions-more-developers-network-egress/). [Concourse issue 6073](https://github.com/concourse/concourse/issues/6073) seeks to develop a plan for dealing with the rate limits, discussion around which mentions perhaps providing Concourse users the ability to configure a global [registry mirror with a pull-through cache](https://docs.docker.com/registry/recipes/mirror/).

However, I don't believe [Concourse issue 6073 discussion](https://github.com/concourse/concourse/issues/6073) currently accounts for the rate limiting that will be imposed on Concourse users of the `oci-build-task` whose `Dockerfile`s leverage `FROM` directives pulling images from `hub.docker.com`. If I understand correctly, `oci-build-task` users would benefit from the ability to configure `oci-build-task` to use a registry mirror with a pull-through cache as well such that, if configured, directives like `FROM foo/bar` would attempt to pull `foo/bar` from the configured registry mirror rather than the rate-limited `hub.docker.com`.

## Context

My team currently operates a rather large Concourse instance serving hundreds of teams and thousands of developers, many of whom use the `oci-build-task` in concert with the [registry-image-resource](github.com/concourse/registry-image-resource/) to build and publish container images. I believe these users would greatly benefit from the ability to configure a registry mirror in their pipelines' use of `oci-build-task`.

### PR DESCRIPTION
**Pull request by jmccann**
Taking a stab at #37

~Haven't tested it for reals yet.  Just verified that it generates config I'd expect.  Working on building an image to use in pipeline to verify now.~

Was able to run a job and intercept it and verify:
1) config generated at location expected
2) config generated as expected
3) traffic went to the specified mirror

Please feel free to take a look and let me know of any opportunities to improve the code.  😄 

fixes #37

### CODE REVIEW

File: buildkitd.go
Comment (jmccann) [inline buildkitd.go:31]: I have this purely to allow overriding for testing ... let me know if you have a better idea on how to allow testing the config file generated.

File: buildkitd_config.go
Comment (jmccann) [inline buildkitd_config.go:21]: We only care about `Registries` for now.  But didn't know if worthwhile to leave all the other structs incase we need to extend ability to write other things to config for buildkitd.

File: types.go
Comment (jmccann) [inline types.go:58]: Let me know if you prefer some other name for specifying mirrors.  I think I've also seen `registry_mirrors` elsewhere.

**Comment by jertel** [comment]
@jmccann Have you heard anything further on this PR? I suspect the urgency for this is climbing quickly right now due to the Docker rate limits being enforced this afternoon.

**Comment by isaacsanders** [comment]
@vito ???

**Comment by vito** [comment]
Sorry, will try to review this soon but I'm in the middle of moving and have to be away for a bit.  Thanks for the PR! If anyone is blocked on this it's probably best to just build a fork for now.

**Comment by jmccann** [comment]
> If anyone is blocked on this it's probably best to just build a fork for now.

yep, that's what I've been doing and seems to have been working for me 😄 

good luck moving!

**Comment by sdeleuze** [comment]
@jmccann Does this PR support authenticated mirrors (which is the case in my company)? If yes how could I specify username and passwords?

**Comment by jmccann** [comment]
> @jmccann Does this PR support authenticated mirrors (which is the case in my company)? If yes how could I specify username and passwords?

I don't think this supports that.  I'm also not sure how to support that based on what seems supported in the buildkitd config.

Not sure if figuring out and supporting that should be part of this PR or a separate issue/PR.

**Comment by sdeleuze** [comment]
@jmccann Not a blocking issue anymore I found a way to move forward using just an authenticated user.

**Comment by sdeleuze** [comment]
@vito FYI it seems we can't have access to an anonymous mirror so we would need authentication support (with or without mirror).

**Comment by sdeleuze** [comment]
@jmccann I finally now have access to an anonymous mirror. Could you please provide a sample `REGISTRY_MIRRORS` configuration? I tried to set with a string or an array of string but I get a `invalid payload: json: unknown field "registry_mirror"` error message. I [created a fork](https://github.com/sdeleuze/oci-build-task), defined `mirror` as a default branch and specified `repository: sdeleuze/oci-build-task` is my pipeline. Is that correct?

**Comment by 3choBoomer** [comment]
@jmccann can you provide a little more information on the expected format of the config?

I have used this in my pipeline config, but the output in my docker build step does not appear to be using the mirror:
`- task: build
        privileged: true
        config:
          platform: linux
          image_resource:
            type: registry-image
            source:
              repository: shipt/oci-build-task
          params:
            REGISTRY_MIRRORS: mirror.grc.io`

output of build step:
`#2 [internal] load build definition from Dockerfile
#2 transferring dockerfile: 1.24kB done
#2 DONE 0.2s

#4 [internal] load metadata for docker.io/library/node:12
#4 DONE 1.0s

#3 [internal] load metadata for docker.io/library/node:12-alpine
#3 DONE 1.0s

#6 [stage-1  1/10] FROM docker.io/library/node:12-alpine@sha256:c8b5faa496b...
#6 resolve docker.io/library/node:12-alpine@sha256:c8b5faa496b3eaaaee30e278abee431ec7dfec03e07ca0fc6ee39e31f36e5021 done
#6 DONE 0.0s

#9 [builder 1/7] FROM docker.io/library/node:12@sha256:3caf9292d6d7e47ee207...
#9 resolve docker.io/library/node:12@sha256:3caf9292d6d7e47ee207e49afada5dc3165d((redacted))c8422201dba7c6547dc883efea done
#9 DONE 0.0s

#5 importing cache manifest from local:8831049952644068551
#5 DONE 0.0s

#11 [internal] load build context
#11 transferring context: 243.01MB 1.5s done
#11 DONE 1.7s

#9 [builder 1/7] FROM docker.io/library/node:12@sha256:3caf9292d6d7e47ee207...
#9 DONE 0.1s

#9 [builder 1/7] FROM docker.io/library/node:12@sha256:3caf9292d6d7e47ee207...`

**Comment by mischcon** [comment]
@3choBoomer 
According to @jmccann findings from https://github.com/vito/oci-build-task/issues/37#issuecomment-712464115 its normal that your mirror does not appear in the output.

I validated that with using
```
params:
  REGISTRY_MIRRORS: your.registry.mirror
```
as params for the `oci-build-task` the configured registry mirror gets used.

If you want to be sure, you could either use `tcpdump` to validate or you could try to use an image that is definitely not (yet) in the registry mirror. If the image is part of the registry mirror after the build, you know for sure that your mirror has been used.

**Comment by sdeleuze** [comment]
That's [exactly what I have ](https://github.com/spring-projects-experimental/spring-graalvm-native/blob/master/ci/pipeline.yml#L233) with my https://github.com/sdeleuze/oci-build-task fork specified [here](https://github.com/spring-projects-experimental/spring-graalvm-native/blob/master/ci/pipeline.yml#L222) but I still have a `invalid payload: json: unknown field "registry_mirror'` error. I am not sure if that matters but we are using Concourse `v6.7.1`.

**Comment by mischcon** [comment]
@sdeleuze Imho the error message is correct, since the param is `registry_mirrors` (json) / `REGISTRY_MIRRORS` (env) (note the trailing 's') as you can see here: https://github.com/sdeleuze/oci-build-task/blob/master/types.go#L58

Just keep in mind that you currently cannot determine whether the mirror was used or not by just comparing the output of the task, you have to either validate with `tcpdump` or check the mirror if the image you pulled is part of the mirror (= if the mirror was used)

**Comment by sdeleuze** [comment]
@mischcon Yeah I have noticed `registry_mirrors` (json) / `REGISTRY_MIRRORS` (env) param from the source code. And I am specifiying `REGISTRY_MIRRORS` in my pipeline configuration. What I don't understand is where `registry_mirror` without a trailing `s` come from. Do you know? Could you elaborate what you mean when you say "the error message is correct"?

To determine if it works, without mirror I have a Docker rate limit error, and with it it should work correctly so that part is straightfoward, but the `registry_mirror` error is a fatal one that stop the build and do not even try to reach the repo as shown [in the logs](https://ci.spring.io/teams/spring-graalvm-native/pipelines/spring-graalvm-native/jobs/build-graalvm-ce-20.3-dev-java11-image/builds/29).

**Comment by 3choBoomer** [comment]
Thanks for the info @mischcon

I still think it would be beneficial to have acceptable configurations documented somewhere.

are all of these valid?
```
params:
  REGISTRY_MIRRORS: your.registry.mirror
```

```
params:
  REGISTRY_MIRRORS: [ "your.registry.mirror" ]
```

```
params:
  REGISTRY_MIRRORS: [ "your.registry.mirror", "your.2nd.registry.mirror" ]
```

**Comment by mischcon** [comment]
> @jmccann I finally now have access to an anonymous mirror. Could you please provide a sample `REGISTRY_MIRRORS` configuration? I tried to set it with a string or an array of string but I get a `invalid payload: json: unknown field "registry_mirror"` error message (without the trailing `s` which is surprising). I [created a fork](https://github.com/sdeleuze/oci-build-task), defined `mirror` as a default branch and specified `repository: sdeleuze/oci-build-task` is my pipeline. Is that correct?

@sdeleuze Sorry, I interpreted your post as 'with the trailing `s` I do not receive an error'.

I tried to reproduce your error, but I was not able to.

I cloned your fork from [here](https://github.com/sdeleuze/oci-build-task), created a Docker image from it and pushed that image to our local artifactory.

My build-task config that I used to reproduce your situation looks like:
```
  - task: build-image
    privileged: true
    config:
      platform: linux
      image_resource:
        type: registry-image
        source:
          repository: artifactory.internal/sdeleuze-oci-build-task
          tag: latest
          username: ((credentials.artifactory-username))
          password: ((credentials.artifactory-password))
      params:
        DOCKERFILE: docker/Dockerfile
        BUILD_ARG_JVMCI_URL: https://github.com/graalvm/graal-jvmci-8/releases/download/jvmci-20.3-b06/openjdk-8u272+10-jvmci-20.3-b06-fastdebug-linux-amd64.tar.gz
        BUILD_ARG_GRAALVM_BRANCH: release/graal-vm/20.3
        REGISTRY_MIRRORS: mirror.artifactory.internal
      inputs:
      - name: docker
      outputs:
      - name: image
      run:
        path: build
```

Output is:
```
selected worker: concourse-worker-1
#1 [internal] load .dockerignore
#1 transferring context: 2B done
#1 DONE 0.1s

#2 [internal] load build definition from Dockerfile
#2 transferring dockerfile: 2.13kB done
#2 DONE 0.2s
...
```

Side note: I am wondering why the [log output](https://ci.spring.io/teams/spring-graalvm-native/pipelines/spring-graalvm-native/jobs/build-graalvm-ce-20.3-dev-java11-image/builds/29) of your `build` task contains 
```
resource script '/opt/resource/check []' failed: exit status 1
```
A thing that just came to my mind (no clue if this is correct) :
Have you configured ["source 'defaults' for resource types"](https://github.com/concourse/concourse/pull/6079)? If so, could it be possible that the error you receive does not have anything to do with the `oci-build-task` itsself but rather with the `registry-image` trying to pull the image?
What I mean is: Usually when executing a task Concourse first tries to find out if there is a newer version of the task docker image available. Afaik this gets done through the `check` command of the `registry-image` resource. If you have misconfigured the global "source 'defaults'", then the check would fail.

**Comment by sdeleuze** [comment]
@mischcon We have indeed defaults configured which are working fine without this PR. I will double check with our IT team what's going on, thanks for your detailed feedback.

I am not a Concourse expert, so sorry if that does not make sense, but it is not 100% clear to me why we could not just improve the plugin to be able to use the `registry_mirror` defined at `registry-image` level like shown in the [release notes](https://github.com/concourse/concourse/releases/tag/v6.7.0) you mentioned (I think that's what we have configured on our instance) rather than introducing a new `REGISTRY_MIRRORS` entry in `params` for that purpose. Couldn't we just reuse `registry_mirror` already potentially defined as a default for `Dockerfile` image retrieval?

**Comment by sdurrheimer** [comment]
Any news on the final solution?
It's becoming really hard to use the oci-build-task with the new docker-hub limit.

**Comment by aranair** [comment]
@sdurrheimer you can work around this for now I think; instead of just running `path: build`, you can 

```
cat <<EOF >> /etc/buildkit/buildkitd.toml
          [registry."docker.io"]
            mirrors = ["xxxx"]
            http = false
            insecure = false
        EOF
```

**Review by vito — APPROVED** [review]
Thanks for your patience, still been busy with the move and didn't find time to give this a proper review. I was thinking about trimming out the unused config stuff until we need it but let's just merge this in and address that all later. The changes look good to me and it sounds like it works, so 👍

Thanks for your PR and thanks for the tests! I'd like to take a swing at higher level black box tests at some point but I think that's too much to ask this late in the review cycle and it'd probably take more exploration. My motivation here is to keep the tests super high level so we can more easily switch to alternative image builders like Kaniko without having to redo all the tests. 

I'm on mobile so I can't ship a new version for this yet, but merging will at least trigger pushing a new image to `vito/oci-build-task:master` so y'all can more easily start using this. I'll see if I can make a few follow-up changes and ship a new version today or tomorrow!

**Pull request merged**
